src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java

```diff
@@ -620,6 +665,69 @@ private String upper(String property, String culture) {
         return property + ".toUpperCase(Locale.forLanguageTag(\"" + culture + "\"))";
     }
 
+    private Tuple<String, String> length(SQLExpr field, String valueName) {
+        String name = nextId("length");
+        if (Strings.isNullOrEmpty(valueName)) {
+            return new Tuple<>(name, def(name, getPropertyOrStringValue(field) + ".length()"));
+        } else {
+            return new Tuple<>(name, getPropertyOrValue(field) + ";" + def(name, valueName + ".length()"));
+        }
+    }
+
+    private Tuple<String, String> replace(SQLExpr field, String target, String replacement, String valueName) {
+        String name = nextId("replace");
+        if (Strings.isNullOrEmpty(valueName)) {
+            return new Tuple<>(name, def(name, getPropertyOrStringValue(field)
+                    + ".replace(" + target + "," + replacement + ")"));
+        } else {
+            return new Tuple<>(name, getPropertyOrStringValue(field) + ";"
+                    + def(name, valueName + ".replace(\" + target + \",\" + replacement + \")"));
+        }
+    }
+
+    // es behavior: both 'start' and return value are 1-index; return 0 if pattern does not exist;
+    // support out-of-bound index
+    private Tuple<String, String> locate(String pattern, SQLExpr source, int start) {
+        String name = nextId("locate");
+        String docSource = getPropertyOrStringValue(source);
+        start = start < 1 ? 0 : start - 1;
+        return new Tuple<>(name, StringUtils.format(
+                "int count=%d;int res=-1;" + "while(count< %s.length() - %s.length()) {"
+                        + "if (%s == %s.substring(count, count+%s.length())) {"
+                        + "res = count;break;} "
+                        + "else {count++;}}"
+                        + def(name, "res+1"),
+                start, docSource, pattern, pattern, docSource, pattern)
+                );
+    }
+
+    private Tuple<String, String> rtrim(SQLExpr field) {
+        String name = nextId("rtrim");
+        String fieldString = getPropertyOrStringValue(field);
+        return new Tuple<>(name, StringUtils.format(
+                "int pos=%s.length();"
+                + "while(%s.substring(pos-1, pos)==' ' && pos > 0) {pos --;} "
+                + def(name, "%s.substring(0, pos)"),
+                fieldString, fieldString, fieldString
+        ));
+    }
+
+    private Tuple<String, String> ltrim(SQLExpr field) {
+        String name = nextId("ltrim");
+        String fieldString = getPropertyOrStringValue(field);
+        return new Tuple<>(name, StringUtils.format(
+                "int pos=0;"
+                + "while(%s.substring(pos, pos+1)==' ' && pos < %s.length()) {pos ++;} "
```

Comment:
From the doc it seems like painless exposes charAt, so you don't need to substrings

https://www.elastic.co/guide/en/elasticsearch/painless/current/painless-api-reference-shared-java-lang.html#painless-api-reference-shared-String